cmd/charm/charmcmd/plugin.go

```diff
@@ -114,53 +114,107 @@ var pluginDescriptionsResults []pluginDescription
 // getPluginDescriptions runs each plugin with "--description".  The calls to
 // the plugins are run in parallel, so the function should only take as long
 // as the longest call.
+// But there was a bug reported on this behavior and it is noticably slow
+// so cache the results and check each plugins mtime to invalidate the plugin.
 func getPluginDescriptions() []pluginDescription {
 	if len(pluginDescriptionsResults) > 0 {
 		return pluginDescriptionsResults
 	}
-	plugins := findPlugins()
+	pluginCacheDir := os.Getenv("HOME") + "/.cache"
+	pluginCache := pluginCacheDir + "/charm-command-cache"
+	plugins, pluginExists := findPlugins()
 	results := []pluginDescription{}
 	if len(plugins) == 0 {
 		return results
 	}
+	if err := os.MkdirAll(pluginCacheDir, os.ModeDir); err != nil {
+		logger.Errorf("creating plugin cache dir: %s, %s", pluginCacheDir, err)
+	}
+	if f, err := os.Open(pluginCache); err == nil {
+		decoder := json.NewDecoder(f)
+		if err = decoder.Decode(&results); err == nil {
+			cachedExists := map[string]int{}
+		compare:
+			for _, cachedPlugin := range results {
+				cachedExists[pluginPrefix+cachedPlugin.Name]++
+				// If a plugin no longer exists in cache, invalidate entire cache.
+				if pluginExists[pluginPrefix+cachedPlugin.Name] == 0 {
+					results = nil
+					break compare
+				}
+				// Compare ModTime of each found plugin to its cached plugin.
+				for _, newp := range plugins {
+					if newp.name == pluginPrefix+cachedPlugin.Name {
+						filename := newp.dir + "/" + newp.name
+						stat, err2 := os.Stat(filename)
+						if err2 != nil {
+							logger.Errorf("could not stat %s", filename, err2)
+							results = nil
+							break compare
+						}
+						mtime, err2 := strconv.ParseInt(cachedPlugin.ModTime, 0, 64)
+						if stat.ModTime().Unix() != mtime {
+							// We could invalidate just this plugin, but it is a rare occurance so invalidate all.
+							results = nil
+							break compare
+						}
+					}
+				}
+
+			}
+			for _, newp := range plugins {
+				// If the cache is missing a found plugin, invalidate entire cache.
+				if cachedExists[newp.name] == 0 {
+					results = nil
+				}
+			}
+			if results != nil {
+				writeResultsCache(pluginCache, results)
```

Comment:
Why is writeResultsCache being called? The only possible modification to results above is setting to nil, which is not this branch?